private/buf/cmd/buf/command/beta/plugin/pluginpush/pluginpush.go

```diff
@@ -109,38 +113,44 @@ func run(
 	}
 	// TODO: Once we support multiple plugin source types, this could be abstracted away
 	// in the bufpluginsource package. This is much simpler for now though.
-	pluginSourceFilePath, err := storage.Exists(ctx, sourceBucket, bufpluginsource.DockerSourceFilePath)
+	dockerfile, err := sourceBucket.Get(ctx, bufpluginsource.DockerSourceFilePath)
 	if err != nil {
-		return bufcli.NewInternalError(err)
+		if errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("please define a %s plugin source file in the target directory", bufpluginsource.DockerSourceFilePath)
+		}
+		return err
 	}
-	if !pluginSourceFilePath {
-		return fmt.Errorf("please define a %s plugin source file in the target directory", bufpluginsource.DockerSourceFilePath)
+	defer func() {
+		retErr = multierr.Append(retErr, dockerfile.Close())
+	}()
+
+	client, err := bufplugindocker.NewClient(bufplugindocker.WithLogger(container.Logger()))
+	if err != nil {
+		return err
 	}
-	// TODO: Build and push the image to the OCI registry with the Docker library.
-	// Once complete, use the digest in RPC below.
-	//
-	// We effectively need to codify the following:
-	//
-	//  $ docker build -f Dockerfile.twirp -t plugins.buf.build/demolab/twirp:v8.1.0-1 .
-	//  $ docker push plugins.buf.build/demolab/twirp:v8.1.0-1
-	//
-	// This will look something along the lines of:
-	//
-	//  readObjectCloser, err := sourceBucket.Get(pluginSourceFilePath)
-	//  if err != nil {
-	//    return nil, err
-	//  }
-	//  defer func() {
-	//    retErr = multierr.Append(retErr, readObjectCloser.Close())
-	//  }()
-	//  dockerfileContent, err := io.ReadAll(readObjectCloser)
-	//  if err != nil {
-	//    return nil, err
-	//  }
-	//  // Build and push to the OCI registry using the Docker library ...
-	//
-	// ---
-	//
+	defer func() {
+		retErr = multierr.Append(retErr, client.Close())
+	}()
+
+	buildResponse, err := client.Build(ctx, dockerfile, pluginConfig, bufplugindocker.BuildParams{
+		ConfigDirPath: container.ConfigDirPath(),
+	})
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if _, err := client.Delete(ctx, buildResponse.Image); err != nil {
```

Comment:
Our Docker image has a unique tag for each built image. We perform a best effort clean up here after a successful build to not clutter up developer's machines with our intermediate images. Buildkit caching doesn't require these (so subsequent builds still use a build cache for speedy builds).

Reply:
Good call - will update!